### PR TITLE
Polish unittests details and execution conditions to adapt to MUSL

### PR DESCRIPTION
**Pull request by chenwhql**
### PR types

Bug fixes
### PR changes

Others
### Describe

Polish unittests details and execution conditions to adapt to MUSL compiling

musl cmake options:
`cmake -DWITH_MUSL=ON -DWITH_CRYPTO=OFF -DWITH_MKL=OFF -DWITH_GPU=OFF -DWITH_TESTING=ON /paddle`

PR changes:
1. fix some unittests cmake conditions error
2. split two unittests to `single test` and `dist test`
3. skip one failed test temporarily

The failed unittests list in musl

| 测试序号 | 模块名称                             | 错误提示摘要                                                 |
| :------- | :----------------------------------- | :----------------------------------------------------------- |
| 49       | test_naive_executor                  | C++ exception with description” C++ Traceback (most recent call last): Error Message Summary: NotFoundError: No variable [a] in current scope. [Hint: var should not be null.] (at /paddle/paddle/fluid/framework/naive_executor.cc:116)” thrown in the test body. |
| 221      | test_avoid_twice_initialization      | File “/paddle/build/python/paddle/fluid/framework.py”, line 1900, in get_op_proto raise ValueError(“Operator \”%s\” has not been registered.” % type) ValueError: Operator “c_broadcast” has not been registered. |
| 333      | test_desc_clone                      | File “/paddle/build/python/paddle/fluid/framework.py”, line 1900, in get_op_proto raise ValueError(“Operator \”%s\” has not been registered.” % type) ValueError: Operator “send” has not been registered. |
| 343      | test_distributed_strategy            | File “/paddle/build/python/paddle/fluid/incubate/fleet/parameter_server/ir/public.py”, line 26, in module from paddle.fluid.core import CommContext ImportError: cannot import name ‘CommContext’ from ‘paddle.fluid.core_avx’ (/paddle/build/python/paddle/fluid/core_avx.so) |
| 396      | test_fake_init_op                    | ERROR: test_fake_init_selected_rows (test_fake_init_op.TestFakeInitOpSelectedRows) File “/paddle/build/python/paddle/fluid/op.py”, line 202, in get_op_info raise ValueError(“The operator: %s is not registered.” % t) ValueError: The operator: fake_init is not registered. |
| 415      | test_fleet_metric                    | File “/paddle/build/python/paddle/fluid/incubate/fleet/parameter_server/ir/public.py”, line 26, in module from paddle.fluid.core import CommContext ImportError: cannot import name ‘CommContext’ from ‘paddle.fluid.core_avx’ (/paddle/build/python/paddle/fluid/core_avx.so) |
| 417      | test_fleet_ps                        | File “/paddle/build/python/paddle/fluid/incubate/fleet/parameter_server/ir/public.py”, line 26, in module from paddle.fluid.core import CommContext ImportError: cannot import name ‘CommContext’ from ‘paddle.fluid.core_avx’ (/paddle/build/python/paddle/fluid/core_avx.so) |
| 418      | test_fleet_pyramid_hash              | File “/paddle/build/python/paddle/fluid/incubate/fleet/parameter_server/ir/public.py”, line 26, in module from paddle.fluid.core import CommContext ImportError: cannot import name ‘CommContext’ from ‘paddle.fluid.core_avx’ (/paddle/build/python/paddle/fluid/core_avx.so) |
| 420      | test_fleet_rolemaker_2               | ERROR: test_pslib_2 (test_fleet_rolemaker_2.TestCloudRoleMaker2) File “/paddle/build/python/paddle/fluid/incubate/fleet/parameter_server/ir/public.py”, line 26, in module from paddle.fluid.core import CommContext ImportError: cannot import name ‘CommContext’ from ‘paddle.fluid.core_avx’ (/paddle/build/python/paddle/fluid/core_avx.so) |
| 424      | test_fleet_rolemaker_new             | ERROR: test_fs_gloo8 (test_fleet_rolemaker_new.TestGlooWithCloudRoleMaker) File “/paddle/build/python/paddle/fluid/incubate/fleet/parameter_server/ir/public.py”, line 26, in module from paddle.fluid.core import CommContext ImportError: cannot import name ‘CommContext’ from ‘paddle.fluid.core_avx’ (/paddle/build/python/paddle/fluid/core_avx.so) not ready endpoints:[‘127.0.0.1:30019’] |
| 428      | test_fleet_utils                     | File “/paddle/build/python/paddle/fluid/incubate/fleet/parameter_server/ir/public.py”, line 26, in module from paddle.fluid.core import CommContext ImportError: cannot import name ‘CommContext’ from ‘paddle.fluid.core_avx’ (/paddle/build/python/paddle/fluid/core_avx.so) |
| 586      | test_lookup_sparse_table_split_op    | ERROR: test_w_is_selected_rows (test_lookup_sparse_table_split_op.TestLookupSpraseTable) File “/paddle/build/python/paddle/fluid/op.py”, line 202, in get_op_info raise ValueError(“The operator: %s is not registered.” % t) ValueError: The operator: lookup_sparse_table_grad_split is not registered. |
| 609      | test_merge_ids_op                    | ERROR: test_check_output (test_merge_ids_op.TestMergeIdsOp) File “/paddle/build/python/paddle/fluid/framework.py”, line 1900, in get_op_proto raise ValueError(“Operator \”%s\” has not been registered.” % type) ValueError: Operator “merge_ids” has not been registered. |
| 628      | test_multiprocess_dataloader_dataset | Cannot use GPU because you have installed CPU version PaddlePaddle. If you want to use GPU, please try to install GPU version PaddlePaddle by: pip install paddlepaddle-gpu If you only have CPU, please change CUDAPlace(0) to be CPUPlace(). |
| 715      | test_program_code                    | ERROR: test_print (test_program_code.TestProgram2Code) ValueError: Operator “listen_and_serv” has not been registered. |
| 750      | test_ref_by_trainer_id_op            | ERROR: test_check_output (test_ref_by_trainer_id_op.TestRefByTrainerIdOp) ValueError: Operator “ref_by_trainer_id” has not been registered. |
| 786      | test_sgd_op                          | W1119 08:59:54.765177 27399 build_strategy.cc:171] fusion_group is not enabled for Windows/MacOS now, and only effective when running with CUDA GPU. W1119 08:59:54.765215 27399 build_strategy.cc:396] fuse_bn_add_act_pass is only supported on GPU, skipped. Child killed |
| 794      | test_sigmoid_focal_loss_op           | FAIL: test_check_grad (test_sigmoid_focal_loss_op.TestSigmoidFocalLossOp1) File “/paddle/build/python/paddle/fluid/tests/unittests/op_test.py”, line 1273, in _assert_is_close self.assertLessEqual(max_diff, max_relative_error, err_msg()) AssertionError: 2.3044646853182973e-07 not less than or equal to 1e-07 : Operator sigmoid_focal_loss error, Gradient Check On CPUPlace variable X (shape: (10, 10), dtype: float64) max gradient diff 2.304465e-07 over limit 1.000000e-07, the first error element is 6, expected 1.169416e-06, but got 1.169416e-06. |
| 807      | test_split_ids_op                    | ERROR: test_check_output (test_split_ids_op.TestSplitIdsOp) ValueError: Operator “split_ids” has not been registered. ERROR: test_check_output (test_split_ids_op.TestSplitSelectedRows) ValueError: The operator: split_ids is not registered. |
| 842      | test_translated_layer                | File “/paddle/build/python/paddle/fluid/dataloader/dataloader*iter.py”, line 560, in \*try_shutdown_all if not self._shutdown: AttributeError: ‘_DataLoaderIterMultiProcess’ object has no attribute ‘_shutdown’ test_translated_layer failed ERROR: test_inference_and_fine_tuning (test_translated_layer.TestTranslatedLayer) File “/usr/local/lib/python3.7/multiprocessing/synchronize.py”, line 59, in __init** unlink_now) OSError: [Errno 24] No file descriptors available |
| 901      | test_auto_checkpoint                 | java_home = os.environ[“JAVA_HOME”] …… KeyError: ‘JAVA_HOME’ /paddle/build/python/paddle/fluid/tests/unittests/dist_test.sh: line 59: hash: nvidia-smi: not found |
| 902      | test_auto_checkpoint1                | java_home = os.environ[“JAVA_HOME”] …… KeyError: ‘JAVA_HOME’ /paddle/build/python/paddle/fluid/tests/unittests/dist_test.sh: line 59: hash: nvidia-smi: not found |
| 903      | test_auto_checkpoint2                | java_home = os.environ[“JAVA_HOME”] …… KeyError: ‘JAVA_HOME’ /paddle/build/python/paddle/fluid/tests/unittests/dist_test.sh: line 59: hash: nvidia-smi: not found |
| 904      | test_auto_checkpoint3                | java_home = os.environ[“JAVA_HOME”] …… KeyError: ‘JAVA_HOME’ /paddle/build/python/paddle/fluid/tests/unittests/dist_test.sh: line 59: hash: nvidia-smi: not found |
| 905      | test_auto_checkpoint_multiple        | java_home = os.environ[“JAVA_HOME”] …… KeyError: ‘JAVA_HOME’ /paddle/build/python/paddle/fluid/tests/unittests/dist_test.sh: line 59: hash: nvidia-smi: not found |
| 906      | test_auto_checkpoint_dist_basic      | java_home = os.environ[“JAVA_HOME”] …… KeyError: ‘JAVA_HOME’ /paddle/build/python/paddle/fluid/tests/unittests/dist_test.sh: line 59: hash: nvidia-smi: not found |
| 907      | test_hdfs1                           | java_home = os.environ[“JAVA_HOME”] …… KeyError: ‘JAVA_HOME’ /paddle/build/python/paddle/fluid/tests/unittests/dist_test.sh: line 59: hash: nvidia-smi: not found |
| 908      | test_hdfs2                           | java_home = os.environ[“JAVA_HOME”] …… KeyError: ‘JAVA_HOME’ /paddle/build/python/paddle/fluid/tests/unittests/dist_test.sh: line 59: hash: nvidia-smi: not found |
| 909      | test_hdfs3                           | java_home = os.environ[“JAVA_HOME”] …… KeyError: ‘JAVA_HOME’ /paddle/build/python/paddle/fluid/tests/unittests/dist_test.sh: line 59: hash: nvidia-smi: not found |
| 981      | test_yolov3                          | FAIL: test_dygraph_static_same_loss (test_yolov3.TestYolov3) Traceback (most recent call last): File “/paddle/build/python/paddle/fluid/tests/unittests/dygraph_to_static/test_yolov3.py”, line 169, in test_dygraph_static_same_loss static_loss)) AssertionError: False is not true : dygraph_loss: [15742.11914062 9392.61047363] static_loss: [nan nan] |

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/CMakeLists.txt [hunk 1]
```diff
@@ -110,6 +112,22 @@ if(WIN32)
     LIST(REMOVE_ITEM TEST_OPS test_ref_by_trainer_id_op)
 endif()
 
+if(NOT WITH_DISTRIBUTE)
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/CMakeLists.txt:121]: `if(WITH_DISTRIBUTE)`?

File: python/paddle/fluid/tests/unittests/test_translated_layer.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ def setUp(self):
             batch_size=BATCH_SIZE,
             shuffle=True,
             drop_last=True,
-            num_workers=2)
+            num_workers=0)
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/test_translated_layer.py:97]: Which UT will cover `num_workers=2`?

File: python/paddle/fluid/tests/unittests/CMakeLists.txt [hunk 3]
```diff
@@ -82,12 +82,14 @@ if(NOT WITH_GPU OR WIN32)
 endif()
 
 if(WIN32)
-    LIST(REMOVE_ITEM TEST_OPS test_boxps)
-    LIST(REMOVE_ITEM TEST_OPS test_trainer_desc)
     LIST(REMOVE_ITEM TEST_OPS test_multiprocess_reader_exception)
+endif()
+
+if(NOT WITH_DISTRIBUTE OR WIN32)
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/CMakeLists.txt]: Some unittests is under WITH_DISTRIBUTE.

File: python/paddle/fluid/tests/unittests/CMakeLists.txt (hunk 1 above)
Comment (chenwhql) [reply python/paddle/fluid/tests/unittests/CMakeLists.txt:121]: if NOT WITH_DISTRIBUTE, these three uts need to be removed

File: python/paddle/fluid/tests/unittests/test_translated_layer.py (hunk 2 above)
Comment (chenwhql) [reply python/paddle/fluid/tests/unittests/test_translated_layer.py:97]: test_multiprocess_dataloader_* related unittests

image

need run exclusive, the multiprocess dataloader is not the test key point of TranslatedLayer

File: python/unittest_py/requirements.txt [hunk 4]
```diff
@@ -6,3 +6,4 @@ opencv-python<=4.2.0.32
 visualdl ; python_version>="3.5"
 paddle2onnx>=0.4
 prettytable
+gym
```
Comment (luotao1) [inline python/unittest_py/requirements.txt]: #29149 have already done it.

File: python/paddle/fluid/tests/unittests/CMakeLists.txt (hunk 3 above)
Comment (chenwhql) [reply python/paddle/fluid/tests/unittests/CMakeLists.txt]: done, thx

File: python/unittest_py/requirements.txt (hunk 4 above)
Comment (chenwhql) [reply python/unittest_py/requirements.txt]: done, thx

**Review by gongweibao — APPROVED** [review]
LGTM for distributed unit tests.

**Review by luotao1 — APPROVED** [review]

**Pull request merged**